tiledb/sm/array_schema/enumeration.cc

```diff
@@ -86,19 +86,42 @@ Enumeration::Enumeration(
     throw EnumerationException("Invalid cell_val_num in Enumeration");
   }
 
-  if (data == nullptr || data_size == 0) {
-    throw EnumerationException("No attribute value data supplied.");
+  // Check input consistency. Non-nullptr inputs need a non-zero size. Nullptr
+  // inputs require a zero data_size.
+  if (data == nullptr && data_size != 0) {
+    throw EnumerationException(
+        "Invalid data input, nullptr with non-zero size.");
+  }
+
+  if (data != nullptr && data_size == 0) {
+    throw EnumerationException(
+        "Invalid data input, non-nullptr with zero size.");
   }
 
-  if (var_size() && (offsets == nullptr || offsets_size == 0)) {
+  if (offsets == nullptr && offsets_size != 0) {
+    throw EnumerationException(
+        "Invalid offsets input, nullptr with non-zero size.");
+  }
+
+  if (offsets != nullptr && offsets_size == 0) {
+    throw EnumerationException(
+        "Invalid offsets input, non-nullptr with zero size.");
+  }
+
```

Comment:
Hmm, I understand the desire to guard against user mistakes, but my understanding is that the C API is a low-level API whose main users (the other language APIs) know what they are doing.

And there are cases you can end up with a non-null pointer and zero size. Imagine you have a vector whose size you cannot know from the start, you `reserve` some elements to avoid frequent resizes, and it turns out that you haven't added anything. Or a non-empty span that gets sliced into becoming empty. Sure you can do `size != 0 ? data : nullptr`, but that's one more gotcha to be aware of.

But I'm not strongly against it, as long as `tiledb_enumeration_alloc` is accordingly documented.

Reply:
> Hmm, I understand the desire to guard against user mistakes, but my understanding is that the C API is a low-level API whose main users (the other language APIs) know what they are doing.

So two things. First, I've run into these issues while using API's I've personally written. And honestly, those are some of the worst. Which leads me to the second point that whenever I design a user input validation API like this, I try to pretend I'm using it after having completely forgotten I've written it. The goal here is to be the polite but firm antithesis of [Postell's Law](https://en.wikipedia.org/wiki/Robustness_principle). "No, you can't do that, but here is why and this is probably how you can do what you want" turns out to be a lot more robust than, "Yolo! I did something vaguely like you asked for!".

However, there are also cases where input can be more commonly "wrong" by some metric even though it turns out to be the more likely metric. For instance, based on your comment I ended up relaxing the checks around var data input because I know lots of folks like to add strings together and then pass the resulting data/size. Given that var sized data is arbitrary and reliant on offsets, I decided to forgo the enforcement of `data == nullptr` when `data_size == 0`. The offsets shows intent on what the user *expected*, so the second check on whether `data` is nullptr or not is a bit redundant.

> But I'm not strongly against it, as long as `tiledb_enumeration_alloc` is accordingly documented.

I'm not sure what you mean here. Or, more specifically, if you'd like to suggest a change there I'm more than open to any suggestions.

Reply:
Filing this so that we can discuss later and make sure Paul is unblocked for his next enumeration item. https://app.shortcut.com/tiledb-inc/story/35810/discuss-allowing-non-null-pointers-with-zero-size-in-capi.